src/manager/checkpoint.rs

```diff
@@ -164,87 +164,103 @@ async fn manage_subnet((child, parent): (Subnet, Subnet), stop_notify: Arc<Notif
+            let submission_epoch = last_exec + period;
+
+            if curr_epoch >= submission_epoch {
+                // First, we check which accounts are in the validator set. This is done by reading
+                // the parent's chain head and requesting the state at that tip set.
+                let parent_head = parent_client.chain_head().await?;
+                // A key assumption we make now is that each block has exactly one tip set. We panic
+                // if this is not the case as it violates our assumption.
+                // TODO: update this logic once the assumption changes (i.e., mainnet)
+                assert_eq!(parent_head.cids.len(), 1);
+                let cid_map = parent_head.cids.first().unwrap().clone();
+                let parent_tip_set = Cid::try_from(cid_map)?;
+
+                let mut validator_set: HashSet<Address, RandomState> = HashSet::new();
+                match subnet_actor_state.validator_set.validators {
+                    None => {}
+                    Some(validators) => {
+                        for v in validators {
+                            validator_set.insert(Address::from_str(v.addr.deref())?);
+                        }
+                    }
+                };
+
+                // Now, for each account defined in the `child` subnet that is in the validator set, we
+                // submit a checkpoint on its behalf.
+                assert_eq!(child_head.cids.len(), 1); // Again, check key assumption
+                let child_tip_set = Cid::try_from(child_head.cids.first().unwrap().clone())?;
+                for account in child.accounts.iter() {
+                    if validator_set.contains(account) {
+                        // check if the validator already voted
+                        let has_voted = parent_client
+                            .ipc_validator_has_voted_bottomup(&child.id, epoch, account)
+                            .await
+                            .map_err(|e| {
+                                log::error!(
+                                    "error checking if validator has voted in subnet: {:?}",
+                                    &child.id
+                                );
+                                e
+                            })?;
+                        if !has_voted {
+                            // submitting the checkpoint synchronously and waiting to be committed.
+                            submit_checkpoint(
+                                child_tip_set,
+                                submission_epoch,
+                                account,
+                                &child,
+                                &child_client,
+                                &parent_client,
+                            )
+                            .await?;
+
+                            // check if by any chance we have the opportunity to submit any outstanding checkpoint we may be
```

Comment:
Is this really necessary? Can't we get away by simply moving forward and submitting a checkpoint for the latest epoch? I'm assuming that by checkpointing some epoch e, we're also implicitly checkpointing all previous epochs.

Reply:
This one is tricky. While theoretically I agree this is the case, practically we don't implement a way to implicitly checkpoint all the "previous history" in the actor because each checkpoint is populated with its own list of bottom-up messages, and we don't know which ones have been executed already in the parent. 

And while I feel that with top-down messages this would be possible, we currently also implement the voting sequentially, where all the checkpoints are expected to be submitted sequentially. Actually, we should double-check that this is the way we are submitting them in top-down checkpoints.